### PR TITLE
Reduce dependence on outside tools

### PR DESCRIPTION
**Pull request by ryanprior**
I wrote this patch to clear out platform- and provider-specific code, simplifying the codebase by getting back to the module's "just a thin wrapper around `vagrant`" roots.

`vagrant-tramp-ssh` is now just 13 lines of readable shell script. `vagrant-tramp.el` depends only on `bash`, `awk`, `vagrant global-status`, and `vagrant ssh`.

### CODE REVIEW

**Comment by ryanprior** [comment]
I've been dogfooding this the past few days and it's behaving well.

The issues I'm motivated to fix are related to the experience of getting set up.

`vagrant global-status` takes about 0.9 seconds on my machine. Due to the round-trip mechanism, this gets called at least twice in the process of making a connection. If you want to work with multiple VMs at once you're going to pay this cost again for each one. We could reduce the load by keeping a local cache of the box info and evicting it when appropriate. We could also work on a `vagrant` patch that makes `global-status` run really fast. I've never hacked on Vagrant so I don't know how realistic that is. `vagrant -v` finishes in 0.2 seconds, so maybe if we can get `global-status` down to the same time scale it would be good enough.

Then there's the mode errors. `vagrant` itself uses a cache for `global-status` info (which makes me wonder why in the world it takes 0.9 seconds to run!) and can report boxes to be running which are not. We could implement an "uh oh" routine which checks `vagrant status` on the nominally running boxen to see which ones actually are, and then evict non-running ones from auto completion.

I'm waiting to see what you think of this patch before I create issues for those and start working on them.

**Comment by dougm** [comment]
Thanks @ryanprior , I'm looking forward to trying this out and reviewing your changes (tomorrow hopefully).  Would be great if @myrjola (issue #15) could take a look too.

**Comment by myrjola** [comment]
Great job @ryanprior! Everything seems to be working just as I want it to work. I think we still could get rid of the `vagrant-tramp-ssh`-script altogether, but otherwise I think this can be merged.

**Comment by ryanprior** [comment]
Great I'm glad it's working! Barring TRAMP features I'm not aware of, getting rid of `vagrant-tramp-ssh` removes our ability to offer semantically meaningful names for completion, which I find unacceptable.

One option looking forward would be to write a patch to TRAMP that allows us to specify a function to transform the domain prior to its being passed to the connect command.

**Comment by myrjola** [comment]
This is probably the easiest implementation for the moment. It will also make this package work for my multi-machine Vagrantfiles. I closed my PR #15 in favor of this implementation. @dougm I have no reservations on getting this merged.

File: vagrant-tramp.el [hunk 1]
```diff
@@ -13,70 +14,111 @@
...
-  (tramp-set-completion-function vagrant-tramp-method
-                                 `((vagrant-tramp-parse ,vagrant-tramp-ssh))))
+                 (tramp-remote-shell-args ("-c")))))
+
+(defconst vagrant-tramp-completion-function-alist
+  '((vagrant-tramp--completions  ""))
+  "Default list of (FUNCTION FILE) pairs to complete vagrant method.")
+
+;;;###autoload
+(with-eval-after-load "tramp"
```
Comment (lunaryorn) [inline vagrant-tramp.el]: You need a dependency on `(emacs "24.4")` in `Package-Requires:` for `with-eval-after-load`, or you need to stick to `eval-after-load` with a quoted `progn`.

Comment (ryanprior) [inline vagrant-tramp.el]: Now it uses eval-after-load.

**Comment by dougm** [comment]
@ryanprior thanks so much, looks and works great for me with Fusion.  And thanks @lunaryorn and @myrjola for the reviews.

**Pull request merged**